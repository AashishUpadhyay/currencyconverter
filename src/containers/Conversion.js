import React from 'react';
import { connect } from 'react-redux';
import FeesTable from '../components/FeesTable';
import * as actions from '../actions/actions';

export class Conversion extends React.Component {
    constructor(props) {
        super(props);
        this.handleOriginAmountChange = this.handleOriginAmountChange.bind(this);
        this.handleDestAmountChange = this.handleDestAmountChange.bind(this);
        this.handleOriginCurrencyChange = this.handleOriginCurrency.bind(this);
        this.handleDestCurrencyChange = this.handleDestinationCurrency.bind(this);
    }

    componentDidMount() {
        this.originAmountInput.focus();
    }

    handleOriginCurrency(event) {
        var newCurrency = event.target.value;

        this.props.dispatch(actions.changeOriginCurrency(newCurrency));

        var payload = {
            originAmount: this.props.originAmount,
            originCurrency: newCurrency,
            destCurrency: this.props.destinationCurrency,
            calcOriginAmount: false
        }

        this.props.dispatch(actions.fetchConversionRate(payload));
        this.props.dispatch(actions.fetchMakeFee(payload));
    }

    handleDestinationCurrency(event) {
        var newCurrency = event.target.value;

        this.props.dispatch(actions.changeDestinationCurrency(newCurrency));

        var payload = {
            destAmount: this.props.originAmount,
            originCurrency: this.props.originCurrency,
            destCurrency: newCurrency,
            calcOriginAmount: true
        }

        this.props.dispatch(actions.fetchConversionRateAndFees(payload));
    }

    handleOriginAmountChange(event) {
        var newAmount = event.target.value;

        // remove unallowed chars
        newAmount = newAmount.replace(',', '')

        // optimistic field updates
        this.props.dispatch(actions.changeOriginAmount(newAmount));

        var payload = {
            originAmount: newAmount,
            originCurrency: this.props.originCurrency,
            destCurrency: this.props.destinationCurrency,
            calcOriginAmount: false
        }

        this.props.dispatch(actions.fetchConversionRate(payload));
        this.props.dispatch(actions.fetchMakeFee(payload));
    }

    handleDestAmountChange(event) {
        var newAmount = event.target.value;

        // remove unallowed chars
        newAmount = newAmount.replace(',', '')

        this.props.dispatch(actions.changeDestinationAmount(newAmount));

        var payload = {
            destAmount: newAmount,
            originCurrency: this.props.originCurrency,
            destCurrency: this.props.destinationCurrency,
            calcOriginAmount: true
        }

        this.props.dispatch(actions.fetchConversionRateAndFees(payload));
    }

    render() {
        if (this.props.errorMsg) {
            var errorMsg = <div className="errorMsg">{this.props.errorMsg}</div>
        }

        return (
            <div>
                {errorMsg}
                <label>Convert</label>&nbsp;
                <input className="amount-field" ref={input => this.originAmountInput = input} onChange={this.handleOriginAmountChange} value={this.props.originAmount} />
                <select value={this.props.originCurrency} onChange={this.handleOriginCurrencyChange}>
                    <option value="USD">USD</option>
                    <option value="EUR">EUR</option>
                    <option value="JPY">JPY</option>
                </select>
                to <input className="amount-field" onChange={this.handleDestAmountChange} value={this.props.destinationAmount} />&nbsp;
                <select value={this.props.destinationCurrency} onChange={this.handleDestCurrencyChange}>
                    <option value="USD">USD</option>
                    <option value="EUR">EUR</option>
                    <option value="JPY">JPY</option>
                </select>


                <br /><br /><br />
                <FeesTable
                    originCurrency={this.props.originCurrency}
                    destinationCurrency={this.props.destinationCurrency}
                    conversionRate={this.props.conversionRate}
                    fee={this.props.feeAmount}
                    total={this.props.totalCost}
                />
            </div>
        )
    }
}

export const mapStateToProps = (state) => {
    return {
        originAmount: state.amount.originAmount,
        destinationAmount: state.amount.destinationAmount,
        conversionRate: state.amount.conversionRate,
        feeAmount: state.amount.feeAmount,
        totalCost: state.amount.totalCost,
        originCurrency: state.amount.originCurrency,
        destinationCurrency: state.amount.destinationCurrency,
        errorMsg: state.error.errorMsg
    }
};

export default connect(mapStateToProps)(Conversion);
