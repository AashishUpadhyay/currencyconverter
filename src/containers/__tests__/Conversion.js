import { Conversion, mapStateToProps } from '../Conversion';
import renderer from 'react-test-renderer';
import React from 'react';

describe('The conversion component tests!', () => {
    describe('Map state to properties!', () => {
        it('it should map the state to props correctly!', () => {
            var amount = {
                originAmount: 100,
                destinationAmount: 120,
                conversionRate: 1.2,
                feeAmount: 10,
                totalCost: 130,
                originCurrency: 'USD',
                destinationCurrency: 'EUR'
            };

            var error = { errorMsg: 'SOME ERROR MSG!' };
            var inputState = { amount: amount, error: error };
            var expectedState = { ...amount, errorMsg: error.errorMsg };
            var mappedState = mapStateToProps(inputState);
            expect(mappedState).toEqual(expectedState);
        })
    });

    describe('Display conversion element!', () => {
        function createNodeMock(element) {
            if (element.type === 'input') {
                return {
                    focus() { },
                };
            }
            return null;
        }

        it('it should not regress', () => {
            const options = { createNodeMock };
            const tree = renderer.create(<Conversion originAmount={100}
                destinationAmount={120}
                conversionRate={1.2}
                feeAmount={10}
                totalCost={130}
                originCurrency={'USD'}
                destinationCurrency={'EUR'}
                errorMsg=''
            />, options).toJSON();
            console.log(tree);
            expect(tree).toMatchSnapshot();
        })
    });
});