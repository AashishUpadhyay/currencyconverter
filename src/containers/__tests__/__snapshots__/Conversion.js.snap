// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The conversion component tests! Display conversion element! it should not regress 1`] = `
<div>
  <label>
    Convert
  </label>
   
  <input
    className="amount-field"
    onChange={[Function]}
    value={100}
  />
  <select
    onChange={[Function]}
    value="USD"
  >
    <option
      value="USD"
    >
      USD
    </option>
    <option
      value="EUR"
    >
      EUR
    </option>
    <option
      value="JPY"
    >
      JPY
    </option>
  </select>
  to 
  <input
    className="amount-field"
    onChange={[Function]}
    value={120}
  />
   
  <select
    onChange={[Function]}
    value="EUR"
  >
    <option
      value="USD"
    >
      USD
    </option>
    <option
      value="EUR"
    >
      EUR
    </option>
    <option
      value="JPY"
    >
      JPY
    </option>
  </select>
  <br />
  <br />
  <br />
  <div>
    <table>
      <tbody>
        <tr>
          <td>
            Conversion Rate
          </td>
          <td>
            1 
            USD
             -&gt; 
            1.20
             
            EUR
          </td>
        </tr>
        <tr>
          <td>
            Fee
          </td>
          <td>
            10.00
             
            USD
          </td>
        </tr>
        <tr>
          <td
            className="total-label"
          >
            Total Cost
          </td>
          <td>
            130.00
             
            USD
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
